lib/src/main/java/io/ably/lib/realtime/ChannelBase.java

```diff
@@ -843,6 +843,12 @@ private void onMessage(final ProtocolMessage protocolMessage) {
             if(msg.connectionId == null) msg.connectionId = protocolMessage.connectionId;
             if(msg.timestamp == 0) msg.timestamp = protocolMessage.timestamp;
             if(msg.id == null) msg.id = protocolMessage.id + ':' + i;
+            // (TM2p)
+            if(msg.version == null) msg.version = String.format("%s:%03d", protocolMessage.channelSerial, i);
+            // (TM2k)
+            if(msg.serial == null) msg.serial = msg.version;
+            // (TM2o)
+            if(msg.createdAt == null) msg.createdAt = msg.timestamp;
```

Comment:
To be pedantic on TM2o - we should only set this if the msg.action is `MESSAGE_CREATE`

Reply:
Fixed for serial as well! Not sure why we make it more complicated by adding this additional condition :)